src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl

```diff
@@ -0,0 +1,171 @@
+    update!(π)
+end
+
+# update policy
+function RLBase.update!(π::MADDPGManager)
+    π.update_step += 1
+    π.update_step % π.update_freq == 0 || return
+
+    for (_, agent) in π.agents
+        length(agent.trajectory) > agent.policy.update_after || return
+        length(agent.trajectory) > π.batch_size || return
+    end
+    
+    # get training data
+    temp_player = rand(keys(π.agents))
+    t = π.agents[temp_player].trajectory
+    inds = rand(π.rng, 1:length(t), π.batch_size)
+    batches = Dict((player, RLCore.fetch!(BatchSampler{SARTS}(π.batch_size), agent.trajectory, inds)) 
+                for (player, agent) in π.agents)
+    
+    # get s, a, s′ for critic
+    s = vcat((batches[player][1] for (player, _) in π.agents)...)
+    a = vcat((batches[player][2] for (player, _) in π.agents)...)
+    s′ = vcat((batches[player][5] for (player, _) in π.agents)...)
+
+    # for training behavior_actor
+    mu_actions = vcat(
+        ((
+            batches[player][1] |> # get personal state information
+            x -> send_to_device(device(agent.policy.behavior_actor), x) |>
+            agent.policy.behavior_actor |> send_to_host
+        ) for (player, agent) in π.agents)...
+    )
+    # for training behavior_critic
+    new_actions = vcat(
+        ((
+            batches[player][5] |> # batch[5] get new_state information
+            x -> send_to_device(device(agent.policy.target_actor), x) |>
+            agent.policy.target_actor |> send_to_host
+        ) for (player, agent) in π.agents)...
+    )
+
+    for (player, agent) in π.agents
+        p = agent.policy
+        A = p.behavior_actor
+        C = p.behavior_critic
+        Aₜ = p.target_actor
+        Cₜ = p.target_critic
+
+        γ = p.γ
+        ρ = p.ρ
+
+        _device(x) = send_to_device(device(A), x)
+
+        # Note that here default A, C, Aₜ, Cₜ on the same device.
+        s, a, s′ = _device((s, a, s′))
+        mu_actions = _device(mu_actions)
+        new_actions = _device(new_actions)
+        r = _device(batches[player][:reward])
+        t = _device(batches[player][:terminal])
+
+        qₜ = Cₜ(vcat(s′, new_actions)) |> vec
+        y = r .+ γ .* (1 .- t) .* qₜ
+
+        gs1 = gradient(Flux.params(C)) do
+            q = C(vcat(s, a)) |> vec
+            loss = mean((y .- q) .^ 2)
+            ignore() do
+                p.critic_loss = loss
+            end
+            loss
+        end
+
+        update!(C, gs1)
+
+        gs2 = gradient(Flux.params(A)) do
+            loss = -mean(C(vcat(s, mu_actions)))
+            ignore() do
+                p.actor_loss = loss
+            end
+            loss
+        end
+
+        update!(A, gs2)
+
+        # polyak averaging
+        for (dest, src) in zip(Flux.params([Aₜ, Cₜ]), Flux.params([A, C]))
+            dest .= ρ .* dest .+ (1 - ρ) .* src
+        end
+
+        s, a, s′ = send_to_host((s, a, s′))
+        mu_actions = send_to_host(mu_actions)
+        new_actions = send_to_host(new_actions)
```

Comment:
Are they required here?

Reply:
Thanks for your kind reviews! I'll check and update my codes later today.